### ISSUE TITLE
feature request: iadmin modresc rebalance visibility

### ISSUE THREAD

**Issue by kript** [irods/irods#3683]
- [x] master
- [x] 4-2-stable
- [x] 4-1-stable
--------

Currently, when a rebalance operation is run on an IES, it is a manual operation. The command provides no feedback while it is running, apart from errors from the logs should it fail. 
in 4.1.11 onwards, logging will appear to show any dirty or single object rebalancing operations that occur. 

However, there is no way for an automated check, another sysadmin, or the iadmin tool itself to check if a rebalance is currently ongoing. The iadmin tool may well happily start a rebalance every time the command is invoked, and it is not clear what behaviour may result. 

This can lead to more than one rebalance being run at one time.

We have also seen some evidence where rebalance command fails, but, as per #3665 may continue to run 'in the background' in some form. 

As a sysadmin, I would like to be able to check if a rebalance was occurring, have a documented way of seeing what a rebalance would do before it was started, ideally have a way of seeing its progress, and a way to safely terminate it. When the rebalance fails, there should be a way to clear up any lock, while retaining the ability to correctly report on whether the rebalance was still running or not.

**Comment by trel** [comment]
How does a metadata element on the resource being rebalanced sound?

A | V | U
-------- | ------- | -------
rebalance_operation | source_hostname | YYYYMMDDHHMMss

Any better suggestions for what to capture in the AVU?

This would allow a new process to check whether the catalog thinks there is a rebalance operation in flight.

**Comment by kript** [comment]
I didn't realise the rebalance worked on a per-resource basis, I thought it was 'just' a list generated by a query it worked its way through, so I've learned something!

Other than that, is the suggestion that to determine if a rebalance is ongoing you check all the resources with imeta ls? Is there a quicker or simpler way to determine if a rebalance is in progress than iterating across all the resources, perhaps with iquest? Apologies - not familiar with using iquest to lookup metadata if its even possible..
It seems a lot of queries to be performed, although presumably only done at the start of a rebalance?

**Comment by trel** [comment]
`iquest` can see the metadata information, so you can run a single query to get the metadata from all the resources that might be running a rebalance:

```
$ imeta add -R demoResc rebalance_operation `hostname`:$$ `date -u +%Y%m%dT%H%M%SZ`
$ imeta add -R r1 rebalance_operation somewhere.nice:$$ `date -u +%Y%m%dT%H%M%SZ`                                                                                                
$ imeta add -R r2 rebalance_operation another.host.name:3456 `date -u +%Y%m%dT%H%M%SZ`                                                                                       

$ iquest "select RESC_NAME, META_RESC_ATTR_VALUE, META_RESC_ATTR_UNITS where META_RESC_ATTR_NAME = 'rebalance_operation'"
RESC_NAME = r2
META_RESC_ATTR_VALUE = another.host.name:3456
META_RESC_ATTR_UNITS = 20180203T142044Z
------------------------------------------------------------
RESC_NAME = demoResc
META_RESC_ATTR_VALUE = x.x.x.x:7294
META_RESC_ATTR_UNITS = 20180203T140006Z
------------------------------------------------------------
RESC_NAME = r1
META_RESC_ATTR_VALUE = somewhere.nice:7294
META_RESC_ATTR_UNITS = 20180203T142006Z
------------------------------------------------------------

$ imeta ls -R demoResc
AVUs defined for resource demoResc:
attribute: rebalance_operation
value: x.x.x.x:7294
units: 20180203T140006Z
```

You can also use the `order()` keyword in genQuery to sort the results:
```
$ iquest "select RESC_NAME, META_RESC_ATTR_VALUE, order(META_RESC_ATTR_UNITS) where META_RESC_ATTR_NAME = 'rebalance_operation'"                                    
RESC_NAME = demoResc
META_RESC_ATTR_VALUE = x.x.x.x:7294
META_RESC_ATTR_UNITS = 20180203T140006Z
------------------------------------------------------------
RESC_NAME = r1
META_RESC_ATTR_VALUE = somewhere.nice:7294
META_RESC_ATTR_UNITS = 20180203T142006Z
------------------------------------------------------------
RESC_NAME = r2
META_RESC_ATTR_VALUE = another.host.name:3456
META_RESC_ATTR_UNITS = 20180203T142044Z
------------------------------------------------------------
```

**Comment by trel** [comment]
I think we're going to stay away from managing the rebalances that are running for now  (`iadmin rebalance start/stop/status`).

Giving visibility into what is running is a start - I think we should see how this works and then later consider how to integrate tools to manage the starting and stopping... if at all.

If we integrate populating this information into the catalog when running `modresc ... rebalance`, we should probably have the same process clean up when that run of the rebalance operation is complete.  But this leaves the possibility of stale information in the catalog (e.g., on a `Ctrl-C` event) to be manually cleaned up later.  This possible stale information should not have any detrimental effects unless/until you build external machinery to care about the values stored in the catalog.  And that's where I think we need to spend some time learning what the real-world use-cases are first.

**Comment by kript** [comment]
Given that a zone with millions of files and a number of faulty checksums, together with the behaviour of rebalance crashing on first issue ( #3524 #3476 #3147 ), having a way to stop a rebalance that may still be going without restarting the service seems rather desirable to me.

I'm also unsure of the proposal above is what we should be able to do now, or proposed for 4.1.12. It appears to be something that you have to do manually; which I'd rather have done automatically rather than have to remember..

**Comment by trel** [comment]
Sorry - didn't mean to confuse anything.

The manual date-based `imeta` command above was proof-of-concept for the data format in the AVU only.... Yes, definitely this type of information would be put into the catalog when `iadmin modresc ... rebalance` is run, and then removed when the rebalance completes successfully.

All the issues you linked to have been closed and are included in 4.1.11 - there should be no known crashing.  And in 4.2, the rebalance even continues for the remaining queued items in the face of an encounter of 'all known replicas are bad'.

I am proposing the inclusion of this automatic 'add AVU on rebalance start' and 'remove AVU on rebalance complete' as part of 4.1.12.

I am proposing any automatic mechanism for restarting/stopping/monitoring these rebalance operations be tabled for now - so we can see some real-world patterns around the AVU addition/removal.  Until an integrated solution is included, the monitoring of ongoing rebalances could be handled via `iquest` (demonstrated above).

**Comment by trel** [comment]
We will insert the timestamp, and then care about it on start.
